File: Backup/PyBluez/AppReciverPyBluez/app/src/main/java/com/example/jasonyin/leosender/MainActivity.java
package com.example.jasonyin.leosender;

import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothServerSocket;
import android.bluetooth.BluetoothSocket;
import android.content.Intent;
import android.os.ParcelUuid;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.Set;
import java.util.UUID;

public class MainActivity extends AppCompatActivity {

    private OutputStream outputStream;
    private InputStream inStream;
    private BluetoothServerSocket mmServerSocket;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

    }
    public void StartGameOne(View view) {
//        try {
//            init();
//            write("python3 game1.py 0");
//            run();
//        }catch (IOException s){
//            Log.e("Error: ", s.toString());
//        }
        AcceptThread();
    }

    public void AcceptThread(){
        BluetoothServerSocket temp = null;
        BluetoothAdapter blueAdapter = BluetoothAdapter.getDefaultAdapter();
        UUID uuid = UUID.fromString("94f39d29-7d6d-437d-973b-fba39e49d4ee");
        try{
            temp = blueAdapter.listenUsingRfcommWithServiceRecord("MyApp", uuid);
            Log.e("UUID: ", uuid.toString());
        } catch (IOException s){
            Log.e("ERROR: ", s.toString());
        }
        mmServerSocket = temp;
        BluetoothSocket socket = null;
        // Keep listening until exception occurs or a socket is returned.
        while (true) {
            try {
                Log.e("SOMETHING: ","Now accepting something");
                socket = mmServerSocket.accept();
            } catch (IOException e) {
                Log.e("ERROR2", "Socket's accept() method failed", e);
                break;
            }

            if (socket != null) {

                // A connection was accepted. Perform work associated with
                // the connection in a separate thread.
                try {
                    inStream = socket.getInputStream();
                    outputStream = socket.getOutputStream();
                    run();
                    Log.e("ERROR3: ", "Boom");
                    socket.close();
                    mmServerSocket.close();
                    break;
                }catch (IOException io){
                    Log.e("ERROR3: ", io.toString());
                }
            }
        }
    }

//    private void init() throws IOException {
//        BluetoothAdapter blueAdapter = BluetoothAdapter.getDefaultAdapter();
//        if (blueAdapter != null) {
//            if (blueAdapter.isEnabled()) {
//                Set<BluetoothDevice> bondedDevices = blueAdapter.getBondedDevices();
//
//                if(bondedDevices.size() > 0) {
//                    Object[] devices = (Object []) bondedDevices.toArray();
//
//                    // printing out the devices
//                    for(Object o : devices){
////                        BluetoothDevice device = (BluetoothDevice) o;
//                        Log.e("CURRENT DEVICE: ", ((BluetoothDevice) o).getName());
//                    }
//                    // devices[0] means the first device
//                    BluetoothDevice device = (BluetoothDevice) devices[0];
//                    ParcelUuid[] uuids = device.getUuids();
//
//                    for(ParcelUuid p : uuids){
//                        Log.e("CURRENT UUIDs: ", p.toString());
//                    }
//
////                    BluetoothSocket socket = device.createRfcommSocketToServiceRecord(uuids[0].getUuid());
//                    UUID uuid = UUID.fromString("94f39d29-7d6d-437d-973b-fba39e49d4ee");
//                    Log.e("CURRENT UUID: ", uuid.toString());
//                    BluetoothSocket socket = device.createRfcommSocketToServiceRecord(uuid);
//                    socket.connect();
//                    outputStream = socket.getOutputStream();
//                    inStream = socket.getInputStream();
//                }
//
//                Log.e("error", "No appropriate paired devices.");
//            } else {
//                Log.e("error", "Bluetooth is disabled.");
//            }
//        }
//    }
//
//    public void write(String s) throws IOException {
//        outputStream.write(s.getBytes());
//    }
//
    public void run() {
        String out = "Hello EV3";
        byte[] build = out.getBytes();
        final int BUFFER_SIZE = 1024;
        byte[] buffer = new byte[BUFFER_SIZE];
        int bytes = 0;
        int b = BUFFER_SIZE;
        /*try {
            Thread.sleep(30000);
            outputStream.write(build);
            outputStream.close();
           Log.e("OUTPUTSTREAM: ", "" + out);

        } catch (IOException e) {
            e.printStackTrace();
        }catch (InterruptedException ie) {
            ie.printStackTrace();
        }*/
        while (true) {
            try {
                bytes = inStream.read(buffer, bytes, BUFFER_SIZE - bytes);
                Log.e("INSTREAM: ", "" + bytes);

            } catch (IOException e) {
                e.printStackTrace();
            }

        }
    }
}

File: LeoApp/app/src/main/java/com/example/akshayc/leoapp/MainActivity.kt
package com.example.akshayc.leoapp

import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import com.parse.Parse
import com.parse.ParseInstallation

import java.io.IOException
import java.io.InputStream
import java.util.UUID

import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothSocket
import android.content.Intent
import android.util.Log
import android.widget.Button
import android.widget.TextView
import java.util.Calendar

class MainActivity : AppCompatActivity() {

    var mmSocket: BluetoothSocket? = null
    var mmDevice: BluetoothDevice? = null

    // use ! to end the stream
    val delimiter: Byte = 33
    var readBufferPosition = 0


    fun runGame(message: String) {
        val uuid = UUID.fromString("94f39d29-7d6d-437d-973b-fba39e49d4ee") //Standard SerialPortService ID
        try {

            mmSocket = mmDevice!!.createRfcommSocketToServiceRecord(uuid)
            if (!mmSocket!!.isConnected) {
                mmSocket!!.connect()
            }
            val msg = message
            val mmOutputStream = mmSocket!!.outputStream
            mmOutputStream.write(msg.toByteArray())

        } catch (e: IOException) {
            e.printStackTrace()
        }
    }



    override fun onCreate(savedInstanceState: Bundle?) {

        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        Parse.initialize(this)
        ParseInstallation.getCurrentInstallation().saveInBackground() //Sets up connection to backend for displau

        //val handler = Handler()

        val firstGame = findViewById<Button>(R.id.game1) as Button
        val secondGame = findViewById<Button>(R.id.game2) as Button
        val thirdGame = findViewById<Button>(R.id.game3) as Button
        val fourthGame = findViewById<Button>(R.id.game4) as Button
        val reaction_min = findViewById<TextView>(R.id.txt_min) as TextView
        val reaction_max = findViewById<TextView>(R.id.txt_max) as TextView
        val reaction_average = findViewById<TextView>(R.id.txt_average) as TextView


        val mBluetoothAdapter = BluetoothAdapter.getDefaultAdapter()

        if (mBluetoothAdapter == null) {
            Log.i("NO CONNECTION", "Bluetooth not supported")
            finish()
        } else {
            class workerThread(private val btMsg: String) : Runnable {

                override fun run() {
                    runGame(btMsg)
                    while (!Thread.currentThread().isInterrupted) {

                        val bytesAvailable: Int
                        var workDone = false

                        try {
                            val mmInputStream: InputStream
                            mmInputStream = mmSocket!!.getInputStream()
                            bytesAvailable = mmInputStream.available()

                            if (bytesAvailable > 0) {
                                val packetBytes = ByteArray(bytesAvailable)
                                Log.e("Raspberry Pi connection", "bytes available")
                                val readBuffer = ByteArray(1024)
                                mmInputStream.read(packetBytes)

//                                Log.e("readBuffer: ", readBuffer.toString());
                                for (i in 0..bytesAvailable - 1) {
                                    val b = packetBytes[i]
                                    if (b == delimiter) {
                                        val encodedBytes = ByteArray(readBufferPosition)
                                        System.arraycopy(readBuffer, 0, encodedBytes, 0, encodedBytes.size)
                                        var data = String(encodedBytes)
                                        readBufferPosition = 0
                                        Log.e("This is our data", data)
                                        //The variable data now contains our full sensor data we can send to our backend
                                        data = data.removePrefix("[")
                                        data = data.removeSuffix("]")

                                        val reflectionList = data.split(",")
                                        val timeSet = mutableListOf<Float>()

                                        for(num in reflectionList){
                                            var number = num
                                            number = number.removeSuffix("\'")
                                            number = number.removePrefix("\'")
                                            timeSet.add(number.toFloat())
                                        }

                                        timeSet.average()
                                        timeSet.min()
                                        timeSet.max()

                                        workDone = true
                                        break

                                    } else {
                                        readBuffer[readBufferPosition++] = b
                                    }
                                }

                                if (workDone == true) {
                                    mmSocket!!.close()
                                    break
                                }

                            }
                        } catch (e: IOException) {
                            e.printStackTrace()
                        }

                    }
                }
            }



            // start game 1 handler

            firstGame.setOnClickListener {

                // Perform action on button click
                var time = Calendar.getInstance().timeInMillis
//                Log.e("Time in milli: ", time.toString())
                Thread(workerThread("game1 " + time.toString())).start()
            }

            secondGame.setOnClickListener {
                // Perform action on button click
                Thread(workerThread("runWheels")).start()
            }

            thirdGame.setOnClickListener {
                // Perform action on button click
                Thread(workerThread("game3")).start()
            }

            fourthGame.setOnClickListener {
                // Perform action on button click
                Thread(workerThread("runFace")).start()
            }

            if (!mBluetoothAdapter.isEnabled) {
                val enableBluetooth = Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE)
                startActivityForResult(enableBluetooth, 0)
            }

            val pairedDevices = mBluetoothAdapter.bondedDevices
            if (pairedDevices.size > 0) {
                for (device in pairedDevices) {
                    if (device.name == "raspberrypi") //Name of our device
                    {
                        Log.d("Connected!", device.name)
                        mmDevice = device
                        break
                    }
                }
            }
        }
    }

}
